diff --git a/l4util/include/elf.h b/l4util/include/elf.h
index 011bc3a6..ca2fbabe 100644
--- a/l4util/include/elf.h
+++ b/l4util/include/elf.h
@@ -887,7 +887,7 @@ static inline ElfW(Phdr) *l4util_elf_phdr(ElfW(Ehdr) *hdr);
 
 /* Implemeantions */
 static inline
-int l4util_elf_check_magic(ElfW(Ehdr) *hdr)
+int l4util_elf_check_magic(ElfW(Ehdr) const *hdr)
 {
   return    hdr->e_ident[EI_MAG0] == ELFMAG0
          && hdr->e_ident[EI_MAG1] == ELFMAG1
@@ -896,7 +896,7 @@ int l4util_elf_check_magic(ElfW(Ehdr) *hdr)
 }
 
 static inline
-int l4util_elf_check_arch(ElfW(Ehdr) *hdr)
+int l4util_elf_check_arch(ElfW(Ehdr) const *hdr)
 {
   return    hdr->e_ident[EI_CLASS] == L4_ARCH_EI_CLASS
          && hdr->e_ident[EI_DATA]  == L4_ARCH_EI_DATA
@@ -904,7 +904,7 @@ int l4util_elf_check_arch(ElfW(Ehdr) *hdr)
 }
 
 static inline
-ElfW(Phdr) *l4util_elf_phdr(ElfW(Ehdr) *hdr)
+ElfW(Phdr) *l4util_elf_phdr(ElfW(Ehdr) const *hdr)
 {
   return (ElfW(Phdr) *)((char *)hdr + hdr->e_phoff);
 }
